terraform.sh

```diff
@@ -58,16 +61,46 @@ function backoff() {
 command="${1:-apply}"
 exitcode=1
 
+mkdir -p "$DIR_CONFIGURATION"
+mkdir -p "$DIR_VARIABLES"
 mkdir -p "$DIR_STATE_IN"
 mkdir -p "$DIR_STATE_OUT"
 
+# live lookup of terraform resources stored in configmaps
+echo "Fetching configmap $NAMESPACE/$TF_CONFIGURATION_CONFIG_MAP_NAME and storing data in $PATH_CONFIGURATION_MAINTF..."
+kubectl \
+  --namespace="$NAMESPACE" \
+  get configmap "$TF_CONFIGURATION_CONFIG_MAP_NAME" \
+  --output="jsonpath={.data['main\.tf']}" \
+  > "$PATH_CONFIGURATION_MAINTF"
+
+echo "Fetching configmap $NAMESPACE/$TF_CONFIGURATION_CONFIG_MAP_NAME and storing data in $PATH_CONFIGURATION_VARIABLESTF..."
+kubectl \
+  --namespace="$NAMESPACE" \
+  get configmap "$TF_CONFIGURATION_CONFIG_MAP_NAME" \
+  --output="jsonpath={.data['variables\.tf']}" \
+  > "$PATH_CONFIGURATION_VARIABLESTF"
+
+echo "Fetching secret $NAMESPACE/$TF_VARIABLES_SECRET_NAME and storing data in $PATH_VARIABLES..."
+kubectl \
+  --namespace="$NAMESPACE" \
+  get secret "$TF_VARIABLES_SECRET_NAME" \
+  --output="jsonpath={.data['terraform\.tfvars']}" | \
+  base64 -d > "$PATH_VARIABLES"
+
+echo "Fetching configmap $NAMESPACE/$TF_STATE_CONFIG_MAP_NAME and storing data in $PATH_STATE_IN..."
+kubectl \
+  --namespace="$NAMESPACE" \
+  get configmap "$TF_STATE_CONFIG_MAP_NAME" \
+  --output="jsonpath={.data['terraform\.tfstate']}" \
+  > "$PATH_STATE_IN"
```

Comment:
If this call fails because for example the Seed api-server is overloaded and returns GOAWAY, terraform.tfstate file will be empty and the script will try to create the main.tf again. I just reproduce it passing non-existing `$TF_STATE_CONFIG_MAP_NAME `.
Is there a way to fail early when any of the kubectls above fail? This would prevent such cases.

Reply:
Hm, actually it should already exit here if the `kubectl` call fails as we are setting the `-e` flag: https://github.com/gardener/terraformer/blob/master/terraform.sh#L1

Testable with this small script:

```bash
#!/bin/bash -e 

kubectl get cm foo > /tmp/bar

echo "content of /tmp/bar: "
cat /tmp/bar
```

Reply:
(compare with `-e` and without `-e`)

Reply:
```
$ k logs bar.infra.tf-apply-9hk8b -f
Fetching configmap shoot--foo--bar/bar.infra.tf-config and storing data in /tf/main.tf...
Fetching configmap shoot--foo--bar/bar.infra.tf-config and storing data in /tf/variables.tf...
Fetching configmap shoot--foo--bar/bar.infra.tf-vars and storing data in /tfvars/terraform.tfvars...
Fetching configmap shoot--foo--bar/bar.infra.tf-state-foo and storing data in /tfstate/terraform.tfstate...
Error from server (NotFound): configmaps "bar.infra.tf-state-baz" not found

Initializing the backend...

Initializing provider plugins...

The following providers do not have any version constraints in configuration,
so the latest version was installed.

To prevent automatic upgrades to new major versions that may contain breaking
changes, it is recommended to add version = "..." constraints to the
corresponding provider blocks in configuration, with the constraint strings
suggested below.

* provider.aws: version = "~> 2.26"

Terraform has been successfully initialized!

You may now begin working with Terraform. Try running "terraform plan" to see
any changes that are required for your infrastructure. All Terraform commands
should now work.

If you ever set or change modules or backend configuration for Terraform,
rerun this command to reinitialize your working directory. If you forget, other
commands will detect it and remind you to do so if necessary.
aws_key_pair.kubernetes: Creating...
aws_iam_role.nodes: Creating...
aws_iam_role.bastions: Creating...
aws_vpc.vpc: Creating...
```

Probably the reason is that the terraform currently is started with `sh -c "sh terraform.sh ..."`

Reply:
Hm, yes, indeed. It should be fixed with #36 right?